File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        
        if obstacleGrid[-1][-1] == 1 or obstacleGrid[0][0]:
            return 0
        
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        dp = []
        for i in range(m):
            dp += [[0]*n]
        
        dp[-1][-1] = 1
        k=n-1
        while k >=0 and obstacleGrid[-1][k] != 1:
            dp[-1][k] = 1
            k-=1
        k=m-1
        while k>=0 and obstacleGrid[k][-1] != 1:
            dp[k][-1] = 1
            k-=1
            
            
        for i in range(m-2, -1, -1):
            for j in range(n-2, -1, -1):
                if obstacleGrid[i][j] != 1:
                    dp[i][j] = dp[i+1][j] + dp[i][j+1]
        
        return dp[0][0]

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {

        int m = obstacleGrid.size(), n = obstacleGrid[0].size();

        if ((obstacleGrid[m - 1][n - 1] == 1) || (obstacleGrid[0][0] == 1)) { return 0; }

        vector<vector<unsigned int>> dp(m, vector<unsigned int>(n, 0));
        dp[m - 1][n - 1] = 1;
        int k = n - 2;
        while (k >= 0 && (obstacleGrid[m - 1][k] != 1)) { 
            dp[m - 1][k] = 1;
            k -= 1;
        }
        k = m - 2;
        while (k >= 0 && (obstacleGrid[k][n-1] != 1)) {
            dp[k][n-1] = 1;
            k -= 1;
        }

        for (int i = m - 2; i >= 0; i--) {
            for (int j = n - 2; j >= 0; j--) {
                if (obstacleGrid[i][j] != 1) { dp[i][j] = dp[i + 1][j] + dp[i][j + 1]; }
            }

        }

        return dp[0][0];

    }
};